### PR TITLE
Creating oauth plugin

### PR DESCRIPTION
**Pull request by alex-seville**
Updating tests to 100% coverage

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -31,11 +31,14 @@
     "eslint": "~2.13.1",
     "eslint-config-flickr": "~1.3.1",
     "istanbul": "~0.4.5",
+    "lolex": "^1.5.1",
     "mocha": "~3.0.2",
     "nock": "~8.0.0",
     "require-dir": "~0.3.1"
   },
   "dependencies": {
+    "crypto": "0.0.3",
```
Comment (jeremyruppel) [inline package.json]: We should use the crypto standard lib instead of this.

File: oauth.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+ * @param {object} requestParams
+ * @returns {Superagent request}
+ */
+function addOAuthSignature (requestParams) {
+
+	var oauthParams = {
+		oauth_consumer_key: requestParams.apiKey,
+		oauth_token:        requestParams.accessToken,
+		oauth_version:      '1.0',
+		oauth_timestamp:    Math.round(Date.now() / 1000),
```
Comment (jeremyruppel) [inline oauth.js]: How did this key spacing pass eslint?

File: oauth.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		oauth_nonce: crypto.createHash('sha1').update(String(Date.now())).digest('hex'),
+		oauth_signature_method: 'HMAC-SHA1'
+	};
+
+	// add query params and then sign
+	return req.query(oauthParams)
+	.query({
+		oauth_signature: encodeURIComponent(
```
Comment (jeremyruppel) [inline oauth.js]: Doesn't superagent take care of encoding this for us?

File: oauth.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+ */
+function createSignature (method, url, apiSecret, tokenSecret, params) {
+
+	var baseString = method.toUpperCase() + '&' + encodeURIComponent(url) + '&',
+	    signingKey = apiSecret + "&" + (tokenSecret || '');
+
+	Object.keys(params).sort().forEach(function(key, i) {
+		baseString += (i > 0) ? encodeURIComponent('&') : '';
+		baseString += encodeURIComponent(key + '=' + params[key]);
+	});
```
Comment (jeremyruppel) [inline oauth.js]: Can we manually "sort" the object and use [node's querystring module][1] instead?

[1]: https://nodejs.org/docs/latest-v0.10.x/api/querystring.html

File: oauth.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+			queryStrings = req.qsRaw[0],
+			requestParams = req.qs,
+			method = req.method,
+			url = req.url;
+
+		queryStrings.split(',').forEach(function (queryString) {
+			var keyValuePair = queryStrings.split('=');
+
+			queryStringObj[keyValuePair[0]] = keyValuePair[1];
+		});
```
Comment (jeremyruppel) [inline oauth.js]: So confused, why do we need this? Also, this has to be a typo: `queryStrings` is used in the function body but the argument is `queryString`.

File: oauth.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+		oauth_consumer_key: requestParams.apiKey,
+		oauth_token:        requestParams.accessToken,
+		oauth_version:      '1.0',
+		oauth_timestamp:    Math.round(Date.now() / 1000),
+		oauth_nonce: crypto.createHash('sha1').update(String(Date.now())).digest('hex'),
+		oauth_signature_method: 'HMAC-SHA1'
+	};
+
+	// add query params and then sign
+	return req.query(oauthParams)
```
Comment (jeremyruppel) [inline oauth.js]: I must be missing something. Where does `req` come from here?

File: oauth.js (hunk 5 above)
Comment (alex-seville) [inline oauth.js]: We need to transfer the string into an object

File: oauth.js (hunk 6 above)
Comment (alex-seville) [inline oauth.js]: yep, added it

File: oauth.js (hunk 3 above)
Comment (alex-seville) [inline oauth.js]: seemingly not

File: oauth.js (hunk 2 above)
Comment (jeremyruppel) [inline oauth.js]: https://travis-ci.org/flickr/flickr-sdk/jobs/167470223#L257-L281

File: oauth.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+ * @param {String} tokenSecret
+ * @param {object} params
+ * @returns {String}
+ */
+function createSignature(method, url, apiSecret, tokenSecret, params) {
+
+	var baseString = method.toUpperCase() + '&' + encodeURIComponent(url) + '&',
+	    signingKey = apiSecret + "&" + (tokenSecret || '');
+
+	Object.keys(params).sort().forEach(function (key, i) {
```
Comment (alex-seville) [inline oauth.js]: Tried replacing this with querystring, but it doesn't solve the sorted order in a good reliable way.  This works and matches what the API expects.

**Comment by jeremyruppel** [comment]
:shipit: 👍

**Pull request merged**